index.js

```diff
@@ -1,14 +1,264 @@
+    if (existingURL) {
+      return res.json({
+        message: 'URL already exists',
+        data: {
+          shortUrl: existingURL.shortUrl,
+        },
+      });
+    }
+
+    // Validate customName if provided
+    if (customName) {
+      if (customName.length < 5) {
+        return res.status(400).json({ error: 'Custom name must be at least 5 characters long.' });
+      }
+
+      const formattedCustomName = customName.replace(/\s+/g, '-'); // Replace spaces with dashes
+      existingURL = await Url.findOne({ customName: formattedCustomName });
+
+      if (existingURL) {
+        return res.status(400).json({ error: 'Custom name already exists.' });
+      }
+
+      const newShortenedURL = new ShortenedURL({
+        customName: formattedCustomName,
+        shortUrl: `https://shortit/${formattedCustomName}`,
+        originalUrl: url,
+      });
+
+      await newShortenedURL.save();
+
+      return res.json({
+        message: 'URL shortened successfully',
+        data: {
+          shortUrl: newShortenedURL.shortUrl,
+        },
+      });
+    }
+
+    // Generate random shortUrl
+    const randomShortUrl = nanoid(5);
+    existingURL = await Url.findOne({ shortUrl: randomShortUrl });
+
+    // Ensure random shortUrl is unique
+    while (existingURL) {
+      randomShortUrl = nanoid(5);
+      existingURL = await Url.findOne({ shortUrl: randomShortUrl });
+    }
+
+    const newShortenedURL = new ShortenedURL({
+      shortUrl: `https://shortit/${randomShortUrl}`,
+      originalUrl: url,
+    });
+
+    await newShortenedURL.save();
+
+    return res.json({
+      message: 'URL shortened successfully',
+      data: {
+        shortUrl: newShortenedURL.shortUrl,
+      },
+    });
+  } catch (error) {
+    console.error(error);
+    res.status(500).json({ error: 'Internal Server Error' });
+  }
+});
+  
+ // @route     GET /api/v1/urls
+// @desc      get all urls
+
+app.get('/api/v1/urls', async (req, res) => {
+  try {
+    const allURLs = await Url.find({}, '-_id -__v'); // Exclude _id and __v fields
+    if (allURLs.length === 0) {
+      return res.json([]);
+    }
+
+    const formattedURLs = allURLs.map((url) => ({
+      id: url.id,
+      customName: url.customName,
+      originalUrl: url.originalUrl,
+      shortUrl: url.shortUrl,
+      createdAt: url.createdAt.toISOString(),
+    }));
+
+    return res.json(formattedURLs);
+  } catch (error) {
+    console.error(error);
+    res.status(500).json({ error: 'Internal Server Error' });
+  }
+});
+
+ // @route     GET /api/v1/urls/:id
+// @desc      get single url
+
+app.get('/api/v1/urls/:id', async (req, res) => {
+  try {
+    const { id } = req.params;
+
+    if (!mongoose.Types.ObjectId.isValid(id)) {
+      return res.status(400).json({ error: 'Invalid ID format.' });
+    }
+
+    const url = await Url.findById(id, '-_id -__v'); // Exclude _id and __v fields
+
+    if (!url) {
+      return res.status(404).json({ error: 'URL not found.' });
+    }
```

Comment:
- Move this two if checks to a middleware
- Don't exclude _id and exclude __v in the model